query-algebrizer/src/clauses/tx_log_api.rs

```diff
@@ -0,0 +1,451 @@
+// Copyright 2018 Mozilla
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may not use
+// this file except in compliance with the License. You may obtain a copy of the
+// License at http://www.apache.org/licenses/LICENSE-2.0
+// Unless required by applicable law or agreed to in writing, software distributed
+// under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+// CONDITIONS OF ANY KIND, either express or implied. See the License for the
+// specific language governing permissions and limitations under the License.
+
+use mentat_core::{
+    TypedValue,
+    ValueType,
+};
+
+use mentat_core::util::Either;
+
+use mentat_query::{
+    Binding,
+    FnArg,
+    PlainSymbol,
+    SrcVar,
+    VariableOrPlaceholder,
+    WhereFn,
+};
+
+use clauses::{
+    ConjoiningClauses,
+};
+
+use errors::{
+    BindingError,
+    ErrorKind,
+    Result,
+};
+
+use types::{
+    Column,
+    ColumnConstraint,
+    DatomsTable,
+    Inequality,
+    QualifiedAlias,
+    QueryValue,
+    SourceAlias,
+    TransactionsColumn,
+    TxLogColumn,
+};
+
+use Known;
+
+impl ConjoiningClauses {
+    pub(crate) fn match_one_tx_argument(&self, operator: &PlainSymbol, index: usize, arg: FnArg) -> Result<Either<i64, QualifiedAlias>> {
+        match arg {
+            FnArg::Variable(v) => {
+                // If it's already bound, then let's expand the variable.
+                match self.bound_value(&v) {
+                    Some(TypedValue::Ref(v)) => Ok(Either::Left(v)),
+                    // TODO: handle instants.
+                    // Some(TypedValue::Instant(v)) => Some(v),
+                    Some(tv) => {
+                        bail!(ErrorKind::InputTypeDisagreement(v.name().clone(), ValueType::Ref, tv.value_type()));
+                    },
+                    None => {
+                        // Regardless of whether we'll be providing a ref later, or the value
```

Comment:
We normally do this by simply applying the constraint, following this approximate pattern:

```
            FnArg::Variable(var) => {
                self.constrain_var_to_type(var.clone(), ValueType::Ref);
                if let Some(TypedValue::Ref(e)) = self.bound_value(&var) {
                    // Incorrect types will be handled by the constraint, above.
                    Ok(QueryValue::Entid(e))
                } else {
                    self.column_bindings
                        .get(&var)
                        .and_then(|cols| cols.first().map(|col| QueryValue::Column(col.clone())))
                        .ok_or_else(|| Error::from_kind(ErrorKind::UnboundVariable(var.name())))
                }
            },
```

If the variable isn't bound, we'll get `UnboundVariable`. If the type is wrong, the pattern will be marked known-empty.